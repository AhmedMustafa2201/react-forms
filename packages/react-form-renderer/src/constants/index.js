export const components = {
  BUTTON: 'button',
  TEXT_FIELD: 'text-field',
  TEXTAREA_FIELD: 'textarea-field',
  FIELD_ARRAY: 'field-array',
  SELECT_COMPONENT: 'select-field',
  CHECKBOX: 'checkbox',
  SUB_FORM: 'sub-form',
  RADIO: 'radio',
  TABS: 'tabs',
  TAB_ITEM: 'tab-item',
  DATE_PICKER: 'date-picker',
  TIME_PICKER: 'time-picker',
  WIZARD: 'wizard',
  SWITCH: 'switch-field',
  TEXTAREA: 'textarea-field',
  SELECT: 'select-field',
  PLAIN_TEXT: 'plain-text',
  INPUT_ADDON_GROUP: 'input-addon-group',
  INPUT_ADDON_BUTTON_GROUP: 'input-addon-button-group',
};
  /** Validator functions placeholder */
export const validators = {
  REQUIRED: 'required-validator',
  /**
     * min length of the input value
     */
  MIN_LENGTH: 'min-length-validator',
  /**
     * max length of the input value
     */
  MAX_LENGTH: 'max-length-validator',
  /**
   * exact length of input value
   */
  EXACT_LENGTH: 'exact-length-validator',
  /**
     * minimum count of fileds in some dynamic list of fields
     */
  MIN_ITEMS_VALIDATOR: 'min-items-validator',
  /**
     * Minimum value of number input
     */
  MIN_NUMBER_VALUE: 'min-number-value',
  /**
     * Maximum value of number inpuy
     */
  MAX_NUMBER_VALUE: 'max-number-value',
  /**
     * Regexp pattern validator
     */
  PATTERN_VALIDATOR: 'pattern-validator',
  /**
   * URL validator
   */
  URL: 'url-validator',
};

export const layoutComponents = {
  FORM_WRAPPER: 'FormWrapper',
  BUTTON: 'Button',
  BUTTON_GROUP: 'ButtonGroup',
  TITLE: 'Title',
  DESCRIPTION: 'Description',
};

/**
 * @enum {String}
 */
export const dataTypes = {
  INTEGER: 'integer',
  FLOAT: 'float',
  NUMBER: 'number',
  BOOLEAN: 'boolean',
  STRING: 'string',
};
