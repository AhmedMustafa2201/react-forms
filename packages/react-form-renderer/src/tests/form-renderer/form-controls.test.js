import React from 'react';
import { Form } from 'react-final-form';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import FormControls from '../../form-renderer/form-controls';
import { layoutComponents } from '../../constants';
import RendererContext from '../../form-renderer/renderer-context';

describe('<FormControls />', () => {
  let initialProps;
  let ContextWrapper;
  let initialLayout;

  beforeEach(() => {
    initialLayout = {
      [layoutComponents.BUTTON]: ({ label, buttonType, ...props }) => <button { ...props }>{ label }</button>,
      [layoutComponents.COL]: ({ children, ...props }) => <div { ...props }>{ children }</div>,
      [layoutComponents.BUTTON_GROUP]: ({ children, ...props }) => <div { ...props }>{ children }</div>,
    };
    ContextWrapper = ({ children, ...props }) => (
      <Form onSubmit={ jest.fn() }>
        { () => (
          <RendererContext.Provider
            value={{ layoutMapper: initialLayout, ...props }}
          >
            { children }
          </RendererContext.Provider>
        ) }
      </Form>
    );
    initialProps = {
      onSubmit: jest.fn(),
      onReset: jest.fn(),
      onCancel: jest.fn(),
      canReset: true,
      pristine: true,
    };
  });

  afterEach(() => {
    initialProps.onCancel.mockReset();
    initialProps.onSubmit.mockReset();
    initialProps.onReset.mockReset();
  });

  it('should render all controls and with default labels', () => {
    const wrapper = mount(
      <ContextWrapper>
        <FormControls { ...initialProps }  disableSubmit/>
      </ContextWrapper>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.find('button').first().simulate('click');
    wrapper.find('button').at(1).simulate('click');
    wrapper.find('button').last().simulate('click');

    expect(initialProps.onSubmit).not.toHaveBeenCalled();
    expect(initialProps.onReset).not.toHaveBeenCalled();
    expect(initialProps.onCancel).toHaveBeenCalled();
  });

  it('should render only submit button', () => {
    const wrapper = mount(
      <ContextWrapper>
        <FormControls { ...initialProps } canReset={ false } onCancel={ undefined } />
      </ContextWrapper>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render buttons in correct order', () => {
    const wrapper = mount(
      <ContextWrapper>
        <FormControls { ...initialProps } canSubmit canReset buttonOrder={ [ 'cancel', 'submit', 'reset' ] } />
      </ContextWrapper>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should add missing buttons if not defined in button order', () => {
    const wrapper = mount(
      <ContextWrapper>
        <FormControls { ...initialProps } canSubmit canReset buttonOrder={ [] } />
      </ContextWrapper>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call cancel with form values', () => {
    const wrapper = mount(
      <ContextWrapper>
        <FormControls { ...initialProps } />
      </ContextWrapper>
    );

    const CANCEL_INDEX = 2;
    const FIELD_NAME = 'xxx';
    const FIELD_VALUE = 'yyy';

    const expectedValues =  {
      [FIELD_NAME]: FIELD_VALUE,
    };

    wrapper.find(Form).instance().form.change(FIELD_NAME, FIELD_VALUE);
    wrapper.update();

    wrapper.find('button').at(CANCEL_INDEX).simulate('click');

    expect(initialProps.onCancel).toHaveBeenCalledWith(expectedValues);
  });
});
