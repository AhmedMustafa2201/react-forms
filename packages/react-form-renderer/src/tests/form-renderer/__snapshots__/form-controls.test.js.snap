// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormControls /> should add missing buttons if not defined in button order 1`] = `
<ContextWrapper>
  <ReactFinalForm
    onSubmit={[MockFunction]}
  >
    <FormControls
      buttonOrder={Array []}
      canReset={true}
      canSubmit={true}
      cancelLabel="Cancel"
      onCancel={[MockFunction]}
      onReset={[MockFunction]}
      onSubmit={[MockFunction]}
      pristine={true}
      resetLabel="Reset"
      submitLabel="Submit"
    >
      <ReactFinalForm(FormSpy)>
        <FormSpy
          reactFinalForm={
            Object {
              "batch": [Function],
              "blur": [Function],
              "change": [Function],
              "destroyOnUnregister": false,
              "focus": [Function],
              "getFieldState": [Function],
              "getRegisteredFields": [Function],
              "getState": [Function],
              "initialize": [Function],
              "isValidationPaused": [Function],
              "mutators": Object {},
              "pauseValidation": [Function],
              "registerField": [Function],
              "reset": [Function],
              "resetFieldState": [Function],
              "resumeValidation": [Function],
              "setConfig": [Function],
              "submit": [Function],
              "subscribe": [Function],
            }
          }
        >
          <Component>
            <div>
              <Component
                key="form-submit"
                label="Submit"
                onClick={[Function]}
                type="button"
                variant="primary"
              >
                <button
                  onClick={[Function]}
                  type="button"
                  variant="primary"
                >
                  Submit
                </button>
              </Component>
              <Component
                disabled={true}
                key="form-reset"
                label="Reset"
                onClick={[Function]}
                type="button"
              >
                <button
                  disabled={true}
                  onClick={[Function]}
                  type="button"
                >
                  Reset
                </button>
              </Component>
              <Component
                key="form-cancel"
                label="Cancel"
                onClick={[Function]}
                type="button"
              >
                <button
                  onClick={[Function]}
                  type="button"
                >
                  Cancel
                </button>
              </Component>
            </div>
          </Component>
        </FormSpy>
      </ReactFinalForm(FormSpy)>
    </FormControls>
  </ReactFinalForm>
</ContextWrapper>
`;

exports[`<FormControls /> should render all controls and with default labels 1`] = `
<ContextWrapper>
  <ReactFinalForm
    onSubmit={[MockFunction]}
  >
    <FormControls
      buttonOrder={
        Array [
          "submit",
          "reset",
          "cancel",
        ]
      }
      canReset={true}
      canSubmit={false}
      cancelLabel="Cancel"
      disableSubmit={true}
      onCancel={[MockFunction]}
      onReset={[MockFunction]}
      onSubmit={[MockFunction]}
      pristine={true}
      resetLabel="Reset"
      submitLabel="Submit"
    >
      <ReactFinalForm(FormSpy)>
        <FormSpy
          reactFinalForm={
            Object {
              "batch": [Function],
              "blur": [Function],
              "change": [Function],
              "destroyOnUnregister": false,
              "focus": [Function],
              "getFieldState": [Function],
              "getRegisteredFields": [Function],
              "getState": [Function],
              "initialize": [Function],
              "isValidationPaused": [Function],
              "mutators": Object {},
              "pauseValidation": [Function],
              "registerField": [Function],
              "reset": [Function],
              "resetFieldState": [Function],
              "resumeValidation": [Function],
              "setConfig": [Function],
              "submit": [Function],
              "subscribe": [Function],
            }
          }
        >
          <Component>
            <div>
              <Component
                disabled={true}
                key="form-submit"
                label="Submit"
                onClick={[Function]}
                type="button"
                variant="primary"
              >
                <button
                  disabled={true}
                  onClick={[Function]}
                  type="button"
                  variant="primary"
                >
                  Submit
                </button>
              </Component>
              <Component
                disabled={true}
                key="form-reset"
                label="Reset"
                onClick={[Function]}
                type="button"
              >
                <button
                  disabled={true}
                  onClick={[Function]}
                  type="button"
                >
                  Reset
                </button>
              </Component>
              <Component
                key="form-cancel"
                label="Cancel"
                onClick={[Function]}
                type="button"
              >
                <button
                  onClick={[Function]}
                  type="button"
                >
                  Cancel
                </button>
              </Component>
            </div>
          </Component>
        </FormSpy>
      </ReactFinalForm(FormSpy)>
    </FormControls>
  </ReactFinalForm>
</ContextWrapper>
`;

exports[`<FormControls /> should render buttons in correct order 1`] = `
<ContextWrapper>
  <ReactFinalForm
    onSubmit={[MockFunction]}
  >
    <FormControls
      buttonOrder={
        Array [
          "cancel",
          "submit",
          "reset",
        ]
      }
      canReset={true}
      canSubmit={true}
      cancelLabel="Cancel"
      onCancel={[MockFunction]}
      onReset={[MockFunction]}
      onSubmit={[MockFunction]}
      pristine={true}
      resetLabel="Reset"
      submitLabel="Submit"
    >
      <ReactFinalForm(FormSpy)>
        <FormSpy
          reactFinalForm={
            Object {
              "batch": [Function],
              "blur": [Function],
              "change": [Function],
              "destroyOnUnregister": false,
              "focus": [Function],
              "getFieldState": [Function],
              "getRegisteredFields": [Function],
              "getState": [Function],
              "initialize": [Function],
              "isValidationPaused": [Function],
              "mutators": Object {},
              "pauseValidation": [Function],
              "registerField": [Function],
              "reset": [Function],
              "resetFieldState": [Function],
              "resumeValidation": [Function],
              "setConfig": [Function],
              "submit": [Function],
              "subscribe": [Function],
            }
          }
        >
          <Component>
            <div>
              <Component
                key="form-cancel"
                label="Cancel"
                onClick={[Function]}
                type="button"
              >
                <button
                  onClick={[Function]}
                  type="button"
                >
                  Cancel
                </button>
              </Component>
              <Component
                key="form-submit"
                label="Submit"
                onClick={[Function]}
                type="button"
                variant="primary"
              >
                <button
                  onClick={[Function]}
                  type="button"
                  variant="primary"
                >
                  Submit
                </button>
              </Component>
              <Component
                disabled={true}
                key="form-reset"
                label="Reset"
                onClick={[Function]}
                type="button"
              >
                <button
                  disabled={true}
                  onClick={[Function]}
                  type="button"
                >
                  Reset
                </button>
              </Component>
            </div>
          </Component>
        </FormSpy>
      </ReactFinalForm(FormSpy)>
    </FormControls>
  </ReactFinalForm>
</ContextWrapper>
`;

exports[`<FormControls /> should render only submit button 1`] = `
<ContextWrapper>
  <ReactFinalForm
    onSubmit={[MockFunction]}
  >
    <FormControls
      buttonOrder={
        Array [
          "submit",
          "reset",
          "cancel",
        ]
      }
      canReset={false}
      canSubmit={false}
      cancelLabel="Cancel"
      onReset={[MockFunction]}
      onSubmit={[MockFunction]}
      pristine={true}
      resetLabel="Reset"
      submitLabel="Submit"
    >
      <ReactFinalForm(FormSpy)>
        <FormSpy
          reactFinalForm={
            Object {
              "batch": [Function],
              "blur": [Function],
              "change": [Function],
              "destroyOnUnregister": false,
              "focus": [Function],
              "getFieldState": [Function],
              "getRegisteredFields": [Function],
              "getState": [Function],
              "initialize": [Function],
              "isValidationPaused": [Function],
              "mutators": Object {},
              "pauseValidation": [Function],
              "registerField": [Function],
              "reset": [Function],
              "resetFieldState": [Function],
              "resumeValidation": [Function],
              "setConfig": [Function],
              "submit": [Function],
              "subscribe": [Function],
            }
          }
        >
          <Component>
            <div>
              <Component
                key="form-submit"
                label="Submit"
                onClick={[Function]}
                type="button"
                variant="primary"
              >
                <button
                  onClick={[Function]}
                  type="button"
                  variant="primary"
                >
                  Submit
                </button>
              </Component>
            </div>
          </Component>
        </FormSpy>
      </ReactFinalForm(FormSpy)>
    </FormControls>
  </ReactFinalForm>
</ContextWrapper>
`;
